### PR TITLE
Feature/purge builds

### PR DESCRIPTION
**Pull request by duylethanh**
This PR contains the  implementation of `simex builds purge` (deletes the `clone_dir`, `compile_dir` and `prefix_dir` of non-dev builds) and `simex develop --purge` (deletes the `compile_dir` and `prefix_dir` of dev builds, if `--delete_source` is used, it will also delete the `source_dir`) .
Both commands take an arbitrary number of build names as positional argument and an arbitrary number of revision names in the `--revision` argument.

Not specifying any build or revision will select all non-dev builds and dev builds respectively.
Only specifying builds will select all the given builds in all revisions.
Only specifying revisions will select all builds in the given revisions.
Specifying both will select the cross-product.

This PR closes https://github.com/hu-macsy/simexpal/issues/38.

### CODE REVIEW

**Comment by duylethanh** [comment]
> Specifying both will select the cross-product.

If a combination does not exist, it will be skipped.

**Comment by duylethanh** [comment]
It is now possible to select builds in the build process when using `simex builds make [-h] [--revisions [REVISIONS [REVISIONS ...]]]
                         [builds [builds ...]]`. (The build selections works as described above)

I also added the command `simex builds remake [-h] [--revisions [REVISIONS [REVISIONS ...]]]
                           [builds [builds ...]]`, which allows a user to rebuild builds (useful for when a user changes the commit-hash in a revision). 

Lastly, simexpal will now compare the commit-hash of the local tag and the commit-hash specified in the `experiments.yml` when building revisions. If they differ, simexpal will update the local tag and pull the commit specified in the `experiments.yml`. This fixes a bug where simexpal would always pull the commit of the local tag even when it differed from the commit-hash in the `experiments.yml`.

The PR is now ready to be merged.

File: simexpal/build.py [hunk 1]
```diff
@@ -203,6 +203,27 @@ def do_step(step_yml, default_workdir=None):
...
+				local_hash = subprocess.check_output(['git', 'rev-parse', generic_tag], cwd=build.repo_dir)
+				local_hash = local_hash.strip().decode()
+
+				if local_hash == git_ref:
+					return False
+				else:
+					remote_refs = subprocess.check_output(['git', 'ls-remote', build.info.git_repo, git_ref]).decode().splitlines()
+
+					if len(remote_refs) == 0:
+						pass
```
Comment (duylethanh) [inline simexpal/build.py:216]: When a remote_ref can not be found, we should raise a `RuntimeError`.

Comment (duylethanh) [reply simexpal/build.py]: Done.

File: simexpal/base.py [hunk 2]
```diff
@@ -750,6 +750,15 @@ def is_compiled(self):
 	def is_installed(self):
 		return os.access(os.path.join(self.prefix_dir, 'installed.simexpal'), os.F_OK)
 
+	def purge(self):
+		if not self.revision.is_dev_build:
+			util.try_rmtree(self.clone_dir)
+		else:
+			util.try_rmtree(self.source_dir)
```
Comment (avdgrinten) [inline simexpal/base.py:757]: This deletes the source directory for dev builds. I think that is quite unexpected.

Comment (duylethanh) [reply simexpal/base.py]: The source dir for dev builds will only be deleted if an extra arg `--delete_source` is used in the purge command

**Comment by duylethanh** [comment]
I made the requested change. PR is ready to be merged

File: scripts/simex [hunk 3]
```diff
@@ -243,41 +278,64 @@ builds_subcmds.required = True
...
 
 dev_builds_parser = main_subcmds.add_parser('develop', help='Build local programs',
-		aliases=['d'], parents=[phase_selection_parser])
+		aliases=['d'], parents=[phase_selection_parser, build_selection_parser])
 dev_builds_parser.set_defaults(cmd=do_develop)
-dev_builds_parser.add_argument('--revision', type=str)
-dev_builds_parser.add_argument('builds', nargs='+', type=str)
+dev_builds_parser.add_argument('--purge', action='store_true', help='Purge selected dev-builds')
+dev_builds_parser.add_argument('-f', action='store_true', help='Confirm action')
+dev_builds_parser.add_argument('--delete_source', action='store_true', help='Deletes the source directory when purging')
```
Comment (avdgrinten) [inline scripts/simex:338]: Parameters should use hyphens, not underscores.

Comment (duylethanh) [reply scripts/simex]: Done

**Review by avdgrinten — APPROVED** [review]
LGTM.

**Pull request merged**